crates/wasmtime/src/func.rs

```diff
@@ -34,6 +37,61 @@ use wasmtime_runtime::{
 /// cloning process only performs a shallow clone, so two cloned `Func`
 /// instances are equivalent in their functionality.
 ///
+/// # `Func` and `async`
+///
+/// Functions from the perspective of WebAssembly are always synchronous. You
+/// might have an `async` function in Rust, however, which you'd like to make
+/// available from WebAssembly. Wasmtime supports asynchronously calling
+/// WebAssembly through native stack switching. You can get some more
+/// information about [asynchronous stores](Store::new_async), but from the
+/// perspective of `Func` it's important to know that whether or not your
+/// [`Store`] is asynchronous will dictate whether you call functions through
+/// [`Func::call`] or [`Func::call_async`] (or the wrappers such as
+/// [`Func::get0`] vs [`Func::get0_async`]).
+///
+/// Note that asynchronous function APIs here are a bit trickier than their
+/// synchronous bretheren. For example [`Func::new_async`] and
+/// [`Func::wrapN_async`](Func::wrap1_async) take explicit state parameters to
+/// allow you to close over the state in the returned future. It's recommended
+/// that you pass state via these parameters instead of through the closure's
+/// environment, which may give Rust lifetime errors. Additionally unlike
+/// synchronous functions which can all get wrapped through [`Func::wrap`]
+/// asynchronous functions need to explicitly wrap based on the number of
+/// parameters that they have (e.g. no wasm parameters gives you
+/// [`Func::wrap0_async`], one wasm parameter you'd use [`Func::wrap1_async`],
+/// etc). Be sure to consult the documentation for [`Func::wrap`] for how the
+/// wasm type signature is inferred from the Rust type signature.
+///
+/// # To `Func::call` or to `Func::getN`
+///
+/// There are four ways to call a `Func`. Half are asynchronus and half are
+/// synchronous, corresponding to the type of store you're using. Within each
+/// half you've got two choices:
+///
+/// * Dynamically typed - if you don't statically know the signature of the
+///   function that you're calling you'll be using [`Func::call`] or
+///   [`Func::call_async`]. These functions take a variable-length slice of
+///   "boxed" arguments in their [`Val`] representation. Additionally the
+///   results are returned as an owned slice of [`Val`]. These methods are the
+///   most heavily optimized due to the dynamic type checks that must occur, in
```

Comment:
> These methods are the most heavily optimized...

Perhaps I'm reading this sentence wrong, but "most heavily optimized" sounds like the opposite intention.